File: day1/gardendecorations/submissions/partially_accepted/worstcasequeries.py
#!/usr/bin/env python3
# based on sl_3logn.py
import sys

# unused imports to test impact on startup performance
from collections import defaultdict
import traceback
import bisect
import math
import heapq
from typing import List
import time


def solve_cycle_rec(n, cycle, parity):
    cn = len(cycle)
    if cn == 1:
        return []
    assert cn > 1
    sched = [
        [[i] for i in range(n)]
        for _ in range(3)
    ]
    ncycle = []
    for i in range(cn//2):
        a = cycle[2*i]
        b = cycle[2*i+1]
        ncycle.append(b)
        if (a > b) ^ parity:
            a, b = b, a
        sched[0][b] = [a, b]
        sched[1][a] = [a, b]
        sched[2][b] = [a, b]
    if cn % 2:
        ncycle.append(cycle[-1])
    sched2 = solve_cycle_rec(n, ncycle, parity^1)
    return sched + sched2


def solve_cycle(cycle):
    # cycle[i] should get value from cycle[i+1]
    n = len(cycle)
    assert sorted(cycle) == list(range(n))
    return solve_cycle_rec(n, cycle, 0)


def solve(perm):
    n = len(perm)
    seen = [False] * n
    total_sched = []
    inv = [-1] * n
    for i in range(n):
        if seen[i]:
            continue
        cycle = []
        at = i
        while not seen[at]:
            cycle.append(at)
            seen[at] = True
            at = perm[at]
        ren = sorted(cycle)
        for j, k in enumerate(ren):
            inv[k] = j
        subsched = solve_cycle([inv[x] for x in cycle])
        while len(total_sched) < len(subsched):
            total_sched.append([[i] for i in range(n)])
        for it, inds in enumerate(subsched):
            tot_inds = total_sched[it]
            for j in range(len(inds)):
                tot_inds[ren[j]] = [ren[x] for x in inds[j]]

    return total_sched


def assert_solution_ok2(perm, schedule, vals):
    n = len(perm)
    target_vals = [vals[perm[i]] for i in range(n)]
    for it, sched in enumerate(schedule):
        nvals = [0] * n
        for j, inds in enumerate(sched):
            x = 0
            for i in inds:
                if it % 2 == 0:
                    assert i <= j
                else:
                    assert i >= j
                x ^= vals[i]
            nvals[j] = x
        vals = nvals
    if vals != target_vals:
        print(vals)
        print(target_vals)
        print(schedule)
    assert vals == target_vals


def assert_solution_ok(perm, schedule):
    n = len(perm)
    import random
    vals = [random.randrange(2**32) for _ in range(n)]
    assert_solution_ok2(perm, schedule, list(range(n)))
    assert_solution_ok2(perm, schedule, vals)


def main():
    iteration, n = map(int, input().split())
    perm = list(map(int, input().split()))
    assert sorted(perm) == list(range(n))

    schedule = solve(perm)
    while len(schedule) < 60:
        schedule.append([[i] for i in range(n)])

    if iteration == 0:
        assert_solution_ok(perm, schedule)
        print(len(schedule))
        sys.exit(0)
    else:
        iteration -= 1
        assert 0 <= iteration < len(schedule)
        sched = schedule[iteration]
        bits = [None] * n
        for j in (range(n) if iteration % 2 == 0 else range(n-1, -1, -1)):
            inds = sched[j]
            bits[j] = int(input())
            x = 0
            for i in inds:
                x ^= bits[i]
            print(x)


if __name__ == "__main__":
    main()

File: day1/gardendecorations/submissions/partially_accepted/worstcasequeries.cpp
// based on sl_3logn.cpp
#include <bits/stdc++.h>
using namespace std;

#define rep(i,f,t) for (int i = f; i < (t); i++)
#define all(x) x.begin(), x.end()
#define trav(a, x) for (auto& a : x)
#define sz(x) (int)(x).size()
typedef vector<int> vi;


vector<vector<vi>> solve_cycle_rec(int n, const vi& cycle, int parity) {
    int cn = sz(cycle);
    if (cn == 1)
        return {};
    assert(cn > 1);
	vector<vi> sc;
	rep(i,0,n) sc.push_back(vi(1, i));
    vector<vector<vi>> sched(3, sc);
    vi ncycle;
	rep(i,0,cn/2) {
        int a = cycle[2*i];
        int b = cycle[2*i+1];
        ncycle.push_back(b);
        if ((a > b) ^ parity)
			swap(a, b);
        sched[0][b] = {a, b};
        sched[1][a] = {a, b};
        sched[2][b] = {a, b};
	}
    if (cn % 2)
        ncycle.push_back(cycle.back());
    auto sched2 = solve_cycle_rec(n, ncycle, parity^1);
	sched.insert(sched.end(), sched2.begin(), sched2.end());
    return sched;
}


vector<vector<vi>> solve_cycle(const vi& cycle) {
    int n = sz(cycle);
	vi co(n);
	rep(i,0,n) assert(cycle[i] >= 0 && cycle[i] < n && !co[cycle[i]]++);
    return solve_cycle_rec(n, cycle, 0);
}


vector<vector<vi>> solve(const vi& perm) {
    int n = sz(perm);
    vi seen(n);
    vector<vector<vi>> total_sched;
    vi inv(n, -1);
	rep(i,0,n) {
		if (seen[i]) continue;
        vi cycle;
        int at = i;
        while (!seen[at]) {
            cycle.push_back(at);
            seen[at] = true;
            at = perm[at];
		}
		vi ren = cycle;
		sort(all(ren));
		rep(j,0,sz(ren))
            inv[ren[j]] = j;
		vi small(sz(ren));
		rep(j,0,sz(ren))
			small[j] = inv[cycle[j]];
        auto subsched = solve_cycle(small);
        while (sz(total_sched) < sz(subsched)) {
			vector<vi> sc;
			rep(i,0,n) sc.push_back({i});
            total_sched.push_back(sc);
		}
		rep(it,0,sz(subsched)) {
			auto& inds = subsched[it];
            auto& tot_inds = total_sched[it];
			rep(j,0,sz(inds)) {
				vi v;
				trav(x, inds[j]) {
					assert(x < sz(ren));
					v.push_back(ren[x]);
				}
                tot_inds[ren[j]] = v;
			}
		}
	}
    return total_sched;
}


void assert_solution_ok2(const vi& perm, const vector<vector<vi>>& schedule, vi vals) {
    int n = sz(perm);
	vi target_vals(n);
	rep(i,0,n) target_vals[i] = vals[perm[i]];
	rep(it,0,sz(schedule)) {
		auto& sched = schedule[it];
		vi nvals(n);
		rep(j,0,sz(sched)) {
			auto& inds = sched[j];
            int x = 0;
			trav(i, inds) {
                if (it % 2 == 0)
                    assert(i <= j);
                else
                    assert(i >= j);
                x ^= vals[i];
			}
            nvals[j] = x;
		}
        vals = nvals;
	}
    assert(vals == target_vals);
}


void assert_solution_ok(const vi& perm, const vector<vector<vi>>& schedule) {
    int n = sz(perm);
	vi vals1(n), vals2(n);
	rep(i,0,n) vals1[i] = i, vals2[i] = rand();
    assert_solution_ok2(perm, schedule, vals1);
    assert_solution_ok2(perm, schedule, vals2);
}


int main() {
    int iteration, n;
	cin >> iteration >> n;
	vi perm(n), co(n);
	rep(i,0,n) {
		cin >> perm[i];
		assert(perm[i] >= 0 && perm[i] < n && !co[perm[i]]++);
	}

    auto schedule = solve(perm);
    while (sz(schedule) < 60) {
		vector<vi> sc;
		rep(i,0,n) sc.push_back({i});
        schedule.push_back(sc);
	}

    if (iteration == 0) {
        assert_solution_ok(perm, schedule);
        cout << sz(schedule) << endl;
		_Exit(0);
	} else {
        iteration--;
        assert(0 <= iteration && iteration < sz(schedule));
        auto& sched = schedule[iteration];
		vi bits(n);
		rep(k,0,n) {
			int j = iteration % 2 == 0 ? k : n-1 - k;
            auto& inds = sched[j];
            cin >> bits[j];
            int x = 0;
			trav(i, inds)
                x ^= bits[i];
			cout << x << endl;
		}
	}
}
